tests/ui/resolve/issue-56593-2.rs

```diff
@@ -0,0 +1,26 @@
+// check-pass
+
+use thing::*;
+
+#[derive(Debug)]
+pub enum Thing {
+    Foo,
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_thing() {
+        let thing = Thing::Foo;
```

Comment:
This code is not tested due `#[cfg(test)]` and `#[test]`, it's better to remove both, they are not needed here.

Reply:
I found it will throw an ambiguity error after remove `#[cfg(test)]`, it's a bug and I should fix it firstly.

Reply:
This problem caused by `use thing::*` arises because `root::Thing`(referencing to `root::thing::Thing`) is defined before expanding `#[derive(Debug)] pub enum Thing { Foo }`. This leads to an ambiguity error when resolve `super::*` in `tests`.  

To solve this issue, I want to introduce a check logic to skip binding as follows: 

```rs
// fn resolve_glob_imports
for (mut key, binding) in bindings {
  // ...
  if import.parent_scope.module.unexpand_invocation.any(|item| item.xxx.ident.name.eq(key.itent.name))  {
    continue ;
  }
  // then try_defined(module, key, binding)
}
```

I want to ask a question: Is there this method correctly? If yes, how can I find the `item.name` from `module.unexpand_inovation`?

Reply:
Which ambiguity is reported exactly, `GlobVsExpanded`? `MoreExpandedVsOuter`?

Reply:
`GlobVsGlob`

Reply:
This error happens after remove `_ if old_binding.is_some() => return t`

For `tests::Foo`, it meets 

```rs
// in clouser of update_resolution
if res != old_binding.res() {
      resolution.binding = Some(this.ambiguity(
          AmbiguityKind::GlobVsGlob,
          old_binding,
          binding,
      ));
  }
```

and `binding` refers to `root::Thing`, `old_binding` refers to `root::thing::Thing`.

Reply:
Could you push the code with this error, and without the fix mentioned in https://github.com/rust-lang/rust/pull/113099#discussion_r1248119357?
I'll check it tomorrow.

Reply:
Added, and I have included `issue-56593-3.rs` for comparison.

Reply:
Let's address all the other comments first, and leave this to the end.
Maybe the issue will resolve itself, like https://github.com/rust-lang/rust/pull/113099#discussion_r1243807278 did.

Reply:
Now let's address this problem: This removal of `_ if old_binding.is_some() => return t` had caused an impact, affecting not only the error in `issue-56593-2.rs` but also the correct behavior in `issue-56593-1.rs` and `issue-105235.rs`.

## Why did `issue-56593-2.rs` work before?

Previously, `tests::Thing` pointed to `root::thing::Thing` because the expansion of the `#[derive(Debug)] enum Thing { Foo }` was postponed until the next stage after resolve `thing::Thing`. This prevented redefinition due to the presence of `_ if old_binding.is_some() => return t`. However, with the removal of this statement, an `GlobVsGlob` error because do `tests::Thing` is defined through `use super::*`,  and it does not know reference `root::Thing` or `root::thing::Thing`.

## How to avoid this problem

I think there are two possible solutions:

### First
Only update the ambiguity error when `old_binding.is_some()`, this code might look like:

```rs
fn update_ambiguity(&self, binding: NameBinding) {
   self.update_resolution(|module,key, (this, resolution)| {
       if let Some(old_binding) = resolution.binding {
           if old_binding.is_glob_import() && binding.is_glob_import() && binding.is_ambiguity() {
                resolution.binding = binding_with_warning_ambiguity;
            }
       }
   });
}

// fn update_resolution
match resolution.binding {
    _ if old_binding.is_some() => {
        for import in module.glob_importers.borrow_mut().iter() {
            let imported_binding = self.import(binding, import);
            let _ = self.update_ambiguity(import.parent_scope.module, key, imported_binding);
        }
        return t
    }
    ...
}
```

This method addresses the fixes for #112713, but does not resolve #56593 and #105235. Furthermore, it may create gaps in the current architecture. But it brings less breaking change.

### Second

Delay the definition of a binding that has the same name in the `unexpand_inovation` under the module. The logic could be as [above comment](https://github.com/rust-lang/rust/pull/113099#discussion_r1248119357)

Reply:
>`GlobVsGlob` error because do `tests::Thing` is defined through `use super::*`,  and it does not know reference `root::Thing` or `root::thing::Thing`.

But there's no glob-vs-glob ambiguity there, `root::Thing` is not a glob.
Why is it considered a glob in `issue-56593-2.rs`, but not in `issue-56593-3.rs`?

Reply:
> Why is it considered a glob in issue-56593-2.rs, but not in issue-56593-3.rs?

Due to the presence of `#[derive(Debug)] `.

In `issue-56593-2.rs`, the `resolve_importers` is executed twice. The first execution occurs before macro expand, where `root::Thing` is defined, referring to `root::thing::Thing`. Then, during the second execution of `resolve_importers` during macro expansion, `root::Thing` is resolved again. 
When updating the binding of `(Mod(root::tests), Key(Thing))`, the `binding`(which comes from `super::*`) pointed to `root::Thing`(which comes from the second execution of `resolve_importers`), while the `old_binding` refers to `root::thing::Thing`(which comes from the first execution of `resolve_importers`). Due to this conflicting situation, ambiguity error occurs. 

But in `issue-56593-3.rs`, all resolutions are completed during the first execution of `resolve_importers`.

Reply:
Ok, I see it now.
When the binding in root changes from `root::thing::Thing` to `root::Thing`, glob importers add a *new* binding to `root::tests` instead of replacing the previous one.

Reply:
Skipping the glob importers part until both `single_imports` and `unresolved_invocations` are empty should be fine, although it's possible that name resolution will stuck and make no progress more often with this scheme, we'll need to test.

In this case we'll still miss glob -> glob updates happening due to `!old_binding.vis.is_at_least(binding.vis, this.tcx)` and `binding.is_ambiguity()` (but it will still be an improvement).

Instead of waiting for `single_imports` / `unresolved_invocations` we could *replace* bindings in glob importers on updates in the source module, it will cover the glob -> glob updates as well.
I'm not sure what effect it will have on some other places returning `Determinacy` values though, like `resolve_ident_in_module_unadjusted_ext`.

Reply:
> replace bindings in glob importers on updates in the source module

This approach causes an ice in `﻿actix-http`, as mentioned in this [comment](https://github.com/rust-lang/rust/pull/113099#issuecomment-1633574396). We need to reconsider this solution.